File: backend/src/main/java/com/project/controller/AdminController.java
package com.project.controller;

import java.time.LocalDateTime;
import java.util.List;

import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.project.dto.admin.request.AdminUpdateDTO;
import com.project.dto.admin.request.ProfessionalVerificationDTO;
import com.project.dto.admin.response.AdminProfessionalViewDTO;
import com.project.dto.admin.response.AdminUserViewDTO;
import com.project.dto.appointment.response.AppointmentResponseDTO;
import com.project.dto.user.request.PasswordUpdateDTO;
import com.project.dto.user.request.UserActivationDTO;
import com.project.dto.user.request.UserUpdateDTO;
import com.project.enums.AppointmentStatus;
import com.project.exception.ApiResponse;
import com.project.exception.ResponseBuilder;
import com.project.security.CustomUserDetails;
import com.project.service.AdminService;
import com.project.service.AppointmentService;
import com.project.service.UserService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/admin")
@RequiredArgsConstructor
public class AdminController {

    private final AppointmentService appointmentService;
    private final AdminService adminService;
    private final UserService userService;

    /* ================= ADMIN SELF ================= */

    @PutMapping("/profile")
    public ResponseEntity<ApiResponse<Object>> updateProfile(
            @RequestBody @Valid AdminUpdateDTO dto,
            Authentication auth) {

        CustomUserDetails user = (CustomUserDetails) auth.getPrincipal();
        adminService.updateOwnProfile(user.getUserId(), dto);

        return ResponseBuilder.success(
                "Profile updated successfully",
                null,
                HttpStatus.NO_CONTENT
        );
    }
    
    @GetMapping("/profile")
    public ResponseEntity<ApiResponse<UserUpdateDTO>> updateProfile(
            Authentication auth) {

        CustomUserDetails user =
                (CustomUserDetails) auth.getPrincipal();

        UserUpdateDTO profile =
                userService.getProfile(user.getUserId());

        return ResponseBuilder.success(
                "Profile found",
                profile,
                HttpStatus.OK
        );
    }

    @PutMapping("/password")
    public ResponseEntity<ApiResponse<Object>> updatePassword(
            @RequestBody @Valid PasswordUpdateDTO dto,
            Authentication auth) {

        CustomUserDetails user = (CustomUserDetails) auth.getPrincipal();
        adminService.updatePassword(user.getUserId(), dto);

        return ResponseBuilder.success(
                "Password updated successfully",
                null,
                HttpStatus.NO_CONTENT
        );
    }

    /* ================= USERS ================= */

    @GetMapping("/users")
    public ResponseEntity<ApiResponse<List<AdminUserViewDTO>>> getUsers() {

        List<AdminUserViewDTO> users = adminService.getAllUsers();

        return ResponseBuilder.success(
                "Users fetched successfully",
                users,
                HttpStatus.OK
        );
    }

    @PutMapping("/users/{id}/activation")
    public ResponseEntity<ApiResponse<Object>> updateUserStatus(
            @PathVariable Long id,
            @RequestBody @Valid UserActivationDTO dto) {

        adminService.updateUserActivation(id, dto.isActive());

        return ResponseBuilder.success(
                "User activation status updated",
                null,
                HttpStatus.OK
        );
    }

    /* ================= PROFESSIONALS ================= */

    @GetMapping("/professionals")
    public ResponseEntity<ApiResponse<List<AdminProfessionalViewDTO>>> getProfessionals() {

        List<AdminProfessionalViewDTO> professionals =
                adminService.getAllProfessionals();

        return ResponseBuilder.success(
                "Professionals fetched successfully",
                professionals,
                HttpStatus.OK
        );
    }

    @PutMapping("/professionals/{userId}/verification")
    public ResponseEntity<ApiResponse<Object>> updateVerification(
            @PathVariable Long userId,
            @RequestBody @Valid ProfessionalVerificationDTO dto) {

        adminService.updateProfessionalVerification(
                userId,
                dto.isVerified()
        );

        return ResponseBuilder.success(
                "Professional verification updated",
                null,
                HttpStatus.OK
        );
    }

    /* ================= APPOINTMENTS ================= */

    @GetMapping("/appointments")
    public ResponseEntity<ApiResponse<List<AppointmentResponseDTO>>> getAll() {

        List<AppointmentResponseDTO> appointments =
                appointmentService.getAllAppointments();

        return ResponseBuilder.success(
                "Appointments fetched successfully",
                appointments,
                HttpStatus.OK
        );
    }

    @GetMapping("/appointments/status/{status}")
    public ResponseEntity<ApiResponse<List<AppointmentResponseDTO>>> byStatus(
            @PathVariable AppointmentStatus status) {

        List<AppointmentResponseDTO> appointments =
                appointmentService.getAppointmentsByStatus(status);

        return ResponseBuilder.success(
                "Appointments fetched by status",
                appointments,
                HttpStatus.OK
        );
    }

    @GetMapping("/appointments/user/{userId}")
    public ResponseEntity<ApiResponse<List<AppointmentResponseDTO>>> byUser(
            @PathVariable Long userId) {

        List<AppointmentResponseDTO> appointments =
                appointmentService.getAppointmentsByUser(userId);

        return ResponseBuilder.success(
                "Appointments fetched by user",
                appointments,
                HttpStatus.OK
        );
    }

    @GetMapping("/appointments/professional/{professionalId}")
    public ResponseEntity<ApiResponse<List<AppointmentResponseDTO>>> byProfessional(
            @PathVariable Long professionalId) {

        List<AppointmentResponseDTO> appointments =
                appointmentService.getAppointmentsByProfessional(professionalId);

        return ResponseBuilder.success(
                "Appointments fetched by professional",
                appointments,
                HttpStatus.OK
        );
    }

    @GetMapping("/appointments/between")
    public ResponseEntity<ApiResponse<List<AppointmentResponseDTO>>> betweenDates(
            @RequestParam
            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
            LocalDateTime start,

            @RequestParam
            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
            LocalDateTime end) {

        List<AppointmentResponseDTO> appointments =
                appointmentService.getAppointmentsBetweenDates(start, end);

        return ResponseBuilder.success(
                "Appointments fetched between dates",
                appointments,
                HttpStatus.OK
        );
    }
}

File: backend_dotnet/SafeMind.API/Controllers/AdminController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using SafeMind.Core.DTOs.Admin.Request;
using SafeMind.Core.DTOs.User.Request;
using SafeMind.Core.Enums;
using SafeMind.Core.Interfaces;
using SafeMind.Core.Responses;
using System.Security.Claims;

namespace SafeMind.API.Controllers
{
    [ApiController]
    [Route("api/admin")]
    [Authorize(Roles = "ROLE_ADMIN")]
    public class AdminController : ControllerBase
    {
        private readonly IAdminService _adminService;
        private readonly IAppointmentService _appointmentService;
        private readonly IUserService _userService;

        public AdminController(IAdminService adminService, IAppointmentService appointmentService, IUserService userService)
        {
            _adminService = adminService;
            _appointmentService = appointmentService;
            _userService = userService;
        }

        private long GetCurrentUserId()
        {
            var userIdClaim = User.FindFirst("userId");
            if (userIdClaim == null) throw new UnauthorizedAccessException("Invalid token");
            return long.Parse(userIdClaim.Value);
        }

        /* ================= ADMIN SELF ================= */

        [HttpPut("profile")]
        public async Task<IActionResult> UpdateProfile([FromBody] AdminUpdateDTO dto)
        {
            try
            {
                var userId = GetCurrentUserId();
                await _adminService.UpdateOwnProfileAsync(userId, dto);
                
                return Ok(ApiResponse<object>.SuccessResponse("Profile updated successfully", null));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpGet("profile")]
        public async Task<IActionResult> GetProfile()
        {
            try
            {
                var userId = GetCurrentUserId();
                var profile = await _userService.GetProfileAsync(userId);
                
                return Ok(ApiResponse<SafeMind.Core.DTOs.User.Request.UserUpdateDTO>.SuccessResponse("Profile found", profile));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpPut("password")]
        public async Task<IActionResult> UpdatePassword([FromBody] PasswordUpdateDTO dto)
        {
            try
            {
                var userId = GetCurrentUserId();
                await _adminService.UpdatePasswordAsync(userId, dto);
                
                return Ok(ApiResponse<object>.SuccessResponse("Password updated successfully", null));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        /* ================= USERS ================= */

        [HttpGet("users")]
        public async Task<IActionResult> GetUsers()
        {
            try
            {
                var users = await _adminService.GetAllUsersAsync();
                
                return Ok(ApiResponse<object>.SuccessResponse("Users fetched successfully", users));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpPut("users/{id}/activation")]
        public async Task<IActionResult> UpdateUserStatus(long id, [FromBody] UserActivationDTO dto)
        {
            try
            {
                await _adminService.UpdateUserActivationAsync(id, dto.Active);
                
                return Ok(ApiResponse<object>.SuccessResponse("User activation status updated", null));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        /* ================= PROFESSIONALS ================= */

        [HttpGet("professionals")]
        public async Task<IActionResult> GetProfessionals()
        {
            try
            {
                var professionals = await _adminService.GetAllProfessionalsAsync();
                
                return Ok(ApiResponse<object>.SuccessResponse("Professionals fetched successfully", professionals));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpPut("professionals/{userId}/verification")]
        public async Task<IActionResult> UpdateVerification(long userId, [FromBody] ProfessionalVerificationDTO dto)
        {
            try
            {
                await _adminService.UpdateProfessionalVerificationAsync(userId, dto.Verified);
                
                return Ok(ApiResponse<object>.SuccessResponse("Professional verification updated", null));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        /* ================= APPOINTMENTS ================= */

        [HttpGet("appointments")]
        public async Task<IActionResult> GetAllAppointments()
        {
            try
            {
                var appointments = await _appointmentService.GetAllAppointmentsAsync();
                
                return Ok(ApiResponse<object>.SuccessResponse("Appointments fetched successfully", appointments));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpGet("appointments/status/{status}")]
        public async Task<IActionResult> GetAppointmentsByStatus(AppointmentStatus status)
        {
            try
            {
                var appointments = await _appointmentService.GetAppointmentsByStatusAsync(status);
                
                return Ok(ApiResponse<object>.SuccessResponse("Appointments fetched by status", appointments));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpGet("appointments/user/{userId}")]
        public async Task<IActionResult> GetAppointmentsByUser(long userId)
        {
            try
            {
                var appointments = await _appointmentService.GetAppointmentsByUserIdAsync(userId);
                
                return Ok(ApiResponse<object>.SuccessResponse("Appointments fetched by user", appointments));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpGet("appointments/professional/{professionalId}")]
        public async Task<IActionResult> GetAppointmentsByProfessional(long professionalId)
        {
            try
            {
                var appointments = await _appointmentService.GetAppointmentsByProfessionalIdAsync(professionalId);
                
                return Ok(ApiResponse<object>.SuccessResponse("Appointments fetched by professional", appointments));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }

        [HttpGet("appointments/between")]
        public async Task<IActionResult> GetAppointmentsBetweenDates([FromQuery] DateTime start, [FromQuery] DateTime end)
        {
            try
            {
                var appointments = await _appointmentService.GetAppointmentsBetweenDatesAsync(start, end);
                
                return Ok(ApiResponse<object>.SuccessResponse("Appointments fetched between dates", appointments));
            }
            catch (Exception ex)
            {
                 return BadRequest(ApiResponse<object>.FailureResponse(ex.Message));
            }
        }
    }
}
